core/src/test/scala/integration/kafka/api/TransactionsExpirationTest.scala

```diff
@@ -128,10 +128,13 @@ class TransactionsExpirationTest extends KafkaServerTestHarness {
     producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, 0, "2", "2", willBeCommitted = false))
     producer.flush()
 
+    TestUtils.waitUntilTrue(() => producerState.nonEmpty, s"Producer IDs for topic1 did not propagate quickly")
+
     // Ensure producer IDs are added.
     val pState = producerState
-    assertEquals(1, pState.size)
-    val oldProducerId = pState(0).producerId
+    assertEquals(1, pState.size, s"No producer visible via admin api")
```

Comment:
I don't think the error message makes sense here since we already wait until the `producerState` is not empty. The error here could be the size is different, not no producer visible via admin api. Does that make sense?

Reply:
> I don't think the error message makes sense here since we already wait until the producerState is not empty.

Because we're calling producerState a second time, we might actually get a 0-size list here (or any size, but a rogue producer making the size >1 would be quite unusual.)

Having the message complain about the size mismatch seems less helpful than a message indicating what the reason for the mismatch actually is.

This assertion is just here to cover a race condition in the test, so I'll remove the race condition and replace this with a more accurate assertion error message.

Reply:
To be clear though, the error should be with regards to the producer state not populating right? Not the admin client? But yeah, the new error message makes sense.